xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-ui/src/main/resources/CKEditor/InlineEditor.xml

```diff
@@ -153,6 +153,94 @@
+
+    function avoidCKAndSaveBarOverlap() {
+      // This callback will be called when the page is scrolled or when the
+      // CKEditor's toolbar position is changed.
+
+      var ckBRC = ckFloatingToolbar.getBoundingClientRect();
+      var saveBRC = saveBar.getBoundingClientRect();
+      var collision = ckBRC.top + ckBRC.height &gt; saveBRC.top;
+      if (saveBar.classList.contains("sticky-buttons-fixed")) {
+          // If XWiki's save bar is sticky, it should not have our custom
+          // padding-top.
+          if (saveBar.style.paddingTop) {
+            saveBar.style.paddingTop = '';
+          }
+
+          if (collision) {
+            // If there is a collision, CKEditor decided to place the toolbar at
+            // the bottom. This is the chosen position if:
+            // - the height of the CKEditor toolbar is not greater than the top of
+            //   the editor relative to the viewport, and
+            // - the height of the CKEditor toolbar is not greater than the
+            //   difference between the bottom of the editor relative to the
+            //   viewport and its height
+            // We put the CKEditor toolbar at the top of the viewport.
+            // Note that CKEditor will try to revert this on the events
+            // previously listed but we will force the position back again
+            // immediately thanks to the mutation observer. This does not
+            // trigger an event that makes CKEditor position the toolbar once
+            // again, so we are not racing with it.
+
+            ckFloatingToolbar.style.position = "fixed";
+            ckFloatingToolbar.style.top = "0px";
+          }
+      } else {
+        // If the XWiki's save bar is not sticky, we add a padding-top
+        // corresponding to the CKEditor toolbar's height if there is a
+        // collision, and we revert to the initial padding-top if there is no
+        // collision.
+        if (collision) {
+          if (!saveBar.style.paddingTop) {
+            saveBar.style.paddingTop = (parseInt(getComputedStyle(saveBar).getPropertyValue('padding-top')) + ckBRC.height) + 'px';
+          }
+        } else {
+          if (saveBar.style.paddingTop) {
+            saveBar.style.paddingTop = '';
+          }
+        }
+      }
+    }
+
+    var ckFloatingToolbar = document.getElementById('cke_' + editor.name);
+    var saveBar = document.getElementsByClassName('inplace-editing-buttons')[0];
+    if (ckFloatingToolbar &amp;&amp; saveBar) {
+      window.addEventListener("scroll", avoidCKAndSaveBarOverlap);
+      var mutationObserver = new MutationObserver(avoidCKAndSaveBarOverlap);
+      mutationObserver.observe(ckFloatingToolbar, {
+        attributes: true,
+        attributeFilter: ['style']
+      });
+
+      return {
+        callback: avoidCKAndSaveBarOverlap,
+        mutationObserver: mutationObserver
+      }
```

Comment:
Another option is to return the ``stopAvoidingCKAndSaveBarOverlap`` function (for this handler). The advantage is that you don't need to expose handler internals (callback and mutation observer). The user will only be able to add and remove the handler.

Reply:
I hesitated to do this. I thought it could be confusing but I've seen this pattern elsewhere (in React I think). I'll do this.

Reply:
done